File: other/binarySearch/ts/binarySearch.ts
// binary search- return index of target
export const search = (arr: number[], target: number): number => {
  let targetIndex = -1,
    left = 0,
    right = arr.length - 1;

  while (left <= right && targetIndex < 0) {
    const mid = Math.floor((left + right) / 2);
    if (target < arr[mid]) {
      right = mid - 1;
    } else if (target > arr[mid]) {
      left = mid + 1;
    } else {
      targetIndex = mid;
    }
  }
  return targetIndex;
};

export const searchRecursive = (
  arr: number[],
  target: number,
  left: number,
  right: number
): number => {
  if (left > right) {
    return -1;
  }
  let targetIndex = -1;
  const mid = Math.floor((left + right) / 2);
  if (target < arr[mid]) {
    targetIndex = searchRecursive(arr, target, left, mid - 1);
  } else if (target > arr[mid]) {
    targetIndex = searchRecursive(arr, target, mid + 1, right);
  } else {
    targetIndex = mid;
  }

  return targetIndex;
};

File: other/binarySearch/c/binarySearch.c
#include <stdio.h>
#include <stdlib.h>

int search(int *arr, int size, int target);

int main()
{
  int arr[] = {1, 4, 6, 7, 9, 10, 15, 18, 20};
  int target = 15;
  int size = sizeof(arr) / sizeof(int);

  int result = search(arr, size, target);
  // expected: 6
  printf("result: %d", result);
  return 0;
}

int search(int *arr, int size, int target)
{
  int targetIdx = -1;
  int left = 0;
  int right = size - 1;
  while (left <= right && targetIdx < 0)
  {
    int midIdx = (left + right) / 2;
    if (target > arr[midIdx])
    {
      left = midIdx + 1;
    }
    else if (target < arr[midIdx])
    {
      right = midIdx - 1;
    }
    else
    {
      targetIdx = midIdx;
    }
  }

  return targetIdx;
}
